### ISSUE TITLE
High-pitched Whine from PC speaker sounds

### ISSUE THREAD

**Issue by OpenRift412** [dosbox-staging/dosbox-staging#619]
**Describe the bug**
As the title suggests. Noticed it when fiddling around with the SETUP.EXE for Heretic, which uses PC speaker sounds when moving through the different options in the menu. It does eventually fade away, but it is a little irritating. Something to note is that in the 0.76 alpha version, the whine fades away much quicker, in as little as a quarter of a second.

**Environment (please complete the following information):**
- Windows 10 20H2
- Relevant hardware info: i7 9750H, RTX 2060, 16 GB RAM

**Version of dosbox-staging:**
All current versions have the issue, but the 0.76 alpha has lessened the issue albeit not entirely.

**Configuration parameters:**
Share the config file(s) you've been using to run the program.

**Additional context**
Add any other context or details about the problem you might found relevant
(e.g. screenshots, other dosbox versions impacted, etc).

**Comment by kcgen** [comment]
Thanks for the report @OpenRift412,

I enabled the PC speaker, set cycles to 25000, and navigated the menus in Heretic v1.7.
- Each menu item results in a short square-wave pulse (listen to attached), however I don't hear any whine. 
- [heretic-setup-flac.zip](https://github.com/dosbox-staging/dosbox-staging/files/5273394/heretic-setup-flac.zip)
- Likewise, the same "chk", "chk", "chk" sounds are heard under SVN - however louder, because SVN doesn't dampen down square waves like staging does.

Can you press ctrl + F6 to record it on your side?

**Comment by dreamer** [comment]
Testing Heretic v1.9 (Steam release):

- dosbox-staging v0.76.0-alpha-745-g92969079 - the same as @kcgen's recording
- DOSBox SVN r4367 - the sound artefact is extremely loud and somewhat obnoxious
- DOSBox 0.74-3 - the same as SVN
- DOSBox-X 0.83.5 - it's as loud as SVN, but the quality seems better? Not sure.

Also, notably, the sound is different when you press enter/esc.

I wonder how it is supposed to sound on real hardware?

Considering it only affects `setup.exe` (and not the actual game, which user is more likely to enjoy via SB emulation), there's also an option of disabling pc speaker emulation completely in .conf file for this game ;)

**Comment by dreamer** [comment]
When testing DOOM shareware installer program I noticed it outputs the same bleeps and boops; installer can be downloaded from here: https://www.doomworld.com/idgames/idstuff/doom/

**Comment by kcgen** [comment]
> I wonder how it is supposed to sound on real hardware?

Given the the wave-forms produced is are a `|_|` square wave lasting 0.00161s, my guess is it would sound very similar on hardware too: a not-too-high-piched ~621 Hz "chk", "chk" taps.

@OpenRift412 -- given we can't reproduce this and there is no regression versus SVN, I suggest we close this.  (Unless you can think of other ways for us to reproduce it)

**Comment by dreamer** [comment]
@kcgen can you test in DOSBox-X and compare? Also, there are 3 sounds: one for pressing arrow, one for pressing enter and one for esc.

**Comment by kcgen** [comment]
Yes; tested those too; similar scenario (and no regression).

The DOSBox-X speaker implementation is often not centered about neutral, and this is the case here - so instead of the square wave going positive and then negative and then back to zero, DOSBox-X goes postitive and back to zero, but without the negative portion of the wave form - so there's a bit less audio and the effect is a shorter and "cleaner" blip effect.

So while this is nice here - the behavior isn't necessarily good for games.

Audio signals, as waveforms, should always be zero-centered because the "peak positive" and "peak negative" should produce maximum amplitudes on both sides.  If you're signal is so massively DC-shifted that the "peak negative" portion of the waveform is actually just touching zero (or neutral), then that half of the audio actually has zero amplitude instead of having peak-negative amplitude... and guess what happens to the "peak-positive" wave-form? It actually had double the amplitide that it should.

If you scroll through the staging speaker recordings, you'll see we always are (roughly) centered about zero: https://github.com/dosbox-staging/dosbox-staging/issues/464#issuecomment-651904100

If you perform -X recordings, many times the signal is significantly DC-shifted.  Now.. maybe that's how real hardware behaved when driving a 60-center clock radio speaker in the back of a physical PC.  But it's not how I want to drive my stereo speaker or my ear-buds :-)

**Comment by NicknineTheEagle** [comment]
Can confirm this, Doom Setup program produces echoing whine noise every time "chk" or "beep" sound plays when navigating the menus. Doesn't happen in upstream.

**Comment by kcgen** [comment]
@NicknineTheEagle  - Can you record it w/ Ctrl + F6 and attach here?

**Comment by NicknineTheEagle** [comment]
Here: [setup_000.zip](https://github.com/dosbox-staging/dosbox-staging/files/5285953/setup_000.zip)

**Comment by kcgen** [comment]
It's possible the DC silencer is winding down a bit too quickly, and either my hearing or my headset doesn't have the frequency response to hear it.

In src/hardware/pcspeaker.cpp, feel free adjust:

#define DC_SILENCER_WAVES   5u
#define DC_SILENCER_WAVE_HZ 30u

Add a couple more waves, and drop the HZ to 20, 15, 10, .. and see if the audible effect disappears for you.

**Comment by NicknineTheEagle** [comment]
I feel like this effect is less pronounced in the recording. How strange.

**Comment by kcgen** [comment]
Yes - I hear zero "post-tck" whine; both inside dosbox and directly in the recording.

@dreamer  and  are I both on Linux - which platform are you on? Perhaps this is some adulteration either at the SDL-level or Windows driver level.

**Comment by NicknineTheEagle** [comment]
Here's another recording, this time done in Audacity from Stereo Mix: [speaker.zip](https://github.com/dosbox-staging/dosbox-staging/files/5285960/speaker.zip)
Here's a screenshot to top it off:
![image](https://user-images.githubusercontent.com/6665256/94326143-f0efec00-ffaa-11ea-997c-708eb477002c.png)

**Comment by NicknineTheEagle** [comment]
And here's a recording from upstream to compare: [speaker_svn.zip](https://github.com/dosbox-staging/dosbox-staging/files/5285965/speaker_svn.zip)
Screenshot:
![image](https://user-images.githubusercontent.com/6665256/94326311-d407e880-ffab-11ea-9b0e-f259d7a31ced.png)

I knew I'm not crazy, there is a difference.

**Comment by kcgen** [comment]
Damn - that's not the wave-form that we're providing SDL though. 

Here's the waveform we're giving SDL (from your own recording; which also matches Linux)

![2020-09-25_17-50](https://user-images.githubusercontent.com/1557255/94326235-96797000-ff57-11ea-9a68-86eda022209d.png)

Something on Window is damaging it post-SDL (some kind of time-based filter that's actually altering the frequency?)

Could you try setting `[mixer] rate = 48000` and `[speaker] pcrate = 48000`? Just to make any HDMI codecs happy and possibly access the stream without resampling?

**Comment by NicknineTheEagle** [comment]
Less noticeable but it's still there: [speaker.zip](https://github.com/dosbox-staging/dosbox-staging/files/5285971/speaker.zip)

**Comment by kcgen** [comment]
T

> And here's a recording from upstream to compare: [speaker_svn.zip](https://github.com/dosbox-staging/dosbox-staging/files/5285965/speaker_svn.zip)
> Screenshot:
> ![image](https://user-images.githubusercontent.com/6665256/94326311-d407e880-ffab-11ea-9b0e-f259d7a31ced.png)
> 
> I knew I'm not crazy, there is a difference.

But that's not a direct recording of SVN - that's post-Windows-filtering. Here's directly from SVN (which you should see w/ Ctrl + F6 under SVN):

![2020-09-25_17-57](https://user-images.githubusercontent.com/1557255/94326435-b6f5fa00-ff58-11ea-9f21-ceecaaa283a8.png)

Upstream is also DC-centered, but they take longer to wind-down the DC offset, as you can see in the middle where it starts to angle upward before another square wave hits.

What it looks like Windows is doing, judging by upstream direct output vs. the post-Windows-filtered is that Windows itself is also DC-correcting the stream. 

Any luck adjusting the DC offset defines in `pcspeaker.cpp` for you?   If we can find a variable that's silent on Windows (amidst whatever filtering its doing) then I'd be happy going with that.  We have to roll w/ the punches, even if it doesn't make sense!

**Comment by kcgen** [comment]
![2020-09-25_18-13](https://user-images.githubusercontent.com/1557255/94326726-d42bc800-ff5a-11ea-9f49-e5b97f6768c1.png)

**Comment by NicknineTheEagle** [comment]
> 
> 
> It's possible the DC silencer is winding down a bit too quickly, and either my hearing or my headset doesn't have the frequency response to hear it.
> 
> In src/hardware/pcspeaker.cpp, feel free adjust:
> 
> #define DC_SILENCER_WAVES 5u
> #define DC_SILENCER_WAVE_HZ 30u
> 
> Add a couple more waves, and drop the HZ to 20, 15, 10, .. and see if the audible effect disappears for you.

Tried playing with the values and it doesn't help, the while doesn't go anywhere.

**Comment by kcgen** [comment]
Does the whine exist in these four blips? 

[speaker-windows-artifacts-remaining.zip](https://github.com/dosbox-staging/dosbox-staging/files/5286000/speaker-windows-artifacts-remaining.zip)

Did some googling about "Windows 10 DC offset" turns up lots of hits. 
Users of RealTek audio manager actually expected it to perform DC offset cancellation. 
- https://www.reddit.com/r/audio/comments/26gnnj/realtek_hd_audio_manager_dc_offset_cancellation/

Is it possible to open your mixer and disable/mute all of the the other lines /except/ for PCM playback?

Another option to try is routing your audio via HDMI (and skip the Realtek codec), or also try using a USB headset (which will have it's own internal audio DAC, and similarly skip the onboard codec). I know this is all drastic; but curious if this this Windows-wide, or just specific to the a onboard codec's "Audio Manager" software package.

**Comment by kcgen** [comment]
If some Windows machines (perhaps those running the RealTek Audio Manager) are going to zero the DC offset, then we no longer have control over the signal because this is downstream from us.  In this case, we should provide an option to disable the DC silencer, and let Windows manage it.

**Comment by kcgen** [comment]
@NicknineTheEagle , another option is to perform our silencing on a slower-scale than the RealTek driver. You could try something like this:

``` diff 
-#define DC_SILENCER_WAVES   5u
-#define DC_SILENCER_WAVE_HZ 30u
+#define DC_SILENCER_WAVES   2u
+#define DC_SILENCER_WAVE_HZ 1u
 
 enum SPKR_MODES { SPKR_OFF, SPKR_ON, SPKR_PIT_OFF, SPKR_PIT_ON };
 
@@ -365,7 +365,7 @@ static void PlayOrFadeout(const uint16_t speaker_movements,
                           size_t requested_samples,
                           int16_t *buffer)
 {
-       constexpr uint8_t grace_time_ms = 100u;
+       constexpr uint8_t grace_time_ms = 30u;
        static_assert(grace_time_ms > 0,
```

Do those settings help?

**Comment by NicknineTheEagle** [comment]
> 
> 
> @NicknineTheEagle, going through your audio codec and manager driver settings, is there anything in there about DC offset, "processing", "filtering", or "enhancing"? Active DC-offset correction might be just one stage in some kind of "audio enhancer" type setting. Not that I'm recommending you disable those as a solution - but it would be interesting to know if it's even possible for Windows/RealTek to produce an unadulterated signal.
> 
> If some Windows machines (perhaps those running the RealTek Audio Manager) are going to zero the DC offset, then we no longer have control over the signal because this is downstream from us. In this case, we should provide an option to disable the DC silencer and let Windows manage it.

Latest Realtek drivers don't have any kind of custom control panel. It's just Windows settings for you.

> 
> 
> @NicknineTheEagle , another option is to perform our silencing on a slower-scale than the RealTek driver. You could try something like this:
> 
> ```diff
> -#define DC_SILENCER_WAVES   5u
> -#define DC_SILENCER_WAVE_HZ 30u
> +#define DC_SILENCER_WAVES   2u
> +#define DC_SILENCER_WAVE_HZ 1u
>  
>  enum SPKR_MODES { SPKR_OFF, SPKR_ON, SPKR_PIT_OFF, SPKR_PIT_ON };
>  
> @@ -365,7 +365,7 @@ static void PlayOrFadeout(const uint16_t speaker_movements,
>                            size_t requested_samples,
>                            int16_t *buffer)
>  {
> -       constexpr uint8_t grace_time_ms = 100u;
> +       constexpr uint8_t grace_time_ms = 30u;
>         static_assert(grace_time_ms > 0,
> ```
> 
> Do those settings help?

The noise remains but the fade is really long now: [speaker.zip](https://github.com/dosbox-staging/dosbox-staging/files/5286602/speaker.zip)

**Comment by kcgen** [comment]
Thanks for all the testing @NicknineTheEagle.

Given we have no way to prevent Windows 10 from, as this [Windows 10 Audacity user](https://forum.audacityteam.org/viewtopic.php?f=46&t=96934) puts it, _"the waves seem to go 'astray' "_ our best course of action is to disable our own DC offset correction.

That still won't prevent Windows 10 from altering the audio signal during actual audio and music sequences as that user points out.  In these cases though, we are helpless; apparently Windows 10 users simply have to accept that their audio is going to be manipulated.

With that said, I'm thinking of a new conf option:

``` inii
[speaker]
neutralize_dc_offset = <bool>
```

Regarding its default value -- should this be disabled by default on Windows platforms? (everyone who has an opinion, please comment. I think we should have this addressed for 0.76)
